include/nbl/builtin/hlsl/spirv_intrinsics/core.hlsl

```diff
@@ -47,199 +69,1272 @@ static const uint32_t3 LocalInvocationId;
+	{
+		vk::ext_execution_mode(spv::ExecutionModeSchedulerTargetFmaxMhzINTEL);
+	}
+
+	void maximallyReconvergesKHR()
+	{
+		vk::ext_execution_mode(spv::ExecutionModeMaximallyReconvergesKHR);
+	}
+
+	void fPFastMathDefault()
+	{
+		vk::ext_execution_mode(spv::ExecutionModeFPFastMathDefault);
+	}
+
+	void streamingInterfaceINTEL()
+	{
+		vk::ext_execution_mode(spv::ExecutionModeStreamingInterfaceINTEL);
+	}
+
+	void registerMapInterfaceINTEL()
+	{
+		vk::ext_execution_mode(spv::ExecutionModeRegisterMapInterfaceINTEL);
+	}
+
+	void namedBarrierCountINTEL()
+	{
+		vk::ext_execution_mode(spv::ExecutionModeNamedBarrierCountINTEL);
+	}
+
+	void maximumRegistersINTEL()
+	{
+		vk::ext_execution_mode(spv::ExecutionModeMaximumRegistersINTEL);
+	}
+
+	void maximumRegistersIdINTEL()
+	{
+		vk::ext_execution_mode(spv::ExecutionModeMaximumRegistersIdINTEL);
+	}
+
+	void namedMaximumRegistersINTEL()
+	{
+		vk::ext_execution_mode(spv::ExecutionModeNamedMaximumRegistersINTEL);
+	}
+}
+
+//! Group Operations
+namespace group_operation
+{
+	static const uint32_t Reduce = 0;
+	static const uint32_t InclusiveScan = 1;
+	static const uint32_t ExclusiveScan = 2;
+	static const uint32_t ClusteredReduce = 3;
+	static const uint32_t PartitionedReduceNV = 6;
+	static const uint32_t PartitionedInclusiveScanNV = 7;
+	static const uint32_t PartitionedExclusiveScanNV = 8;
+}
+
+//! Instructions
+template<typename T, typename P>
+[[vk::ext_instruction(spv::OpLoad)]]
+enable_if_t<is_spirv_type_v<P>, T> load(P pointer, [[vk::ext_literal]] uint32_t memoryAccess);
+
+template<typename T, typename P>
+[[vk::ext_instruction(spv::OpLoad)]]
+enable_if_t<is_spirv_type_v<P>, T> load(P pointer, [[vk::ext_literal]] uint32_t memoryAccess, [[vk::ext_literal]] uint32_t memoryAccessParam);
+
+template<typename T, typename P, uint32_t alignment>
+[[vk::ext_instruction(spv::OpLoad)]]
+enable_if_t<is_spirv_type_v<P>, T> load(P pointer, [[vk::ext_literal]] uint32_t __aligned = /*Aligned*/0x00000002, [[vk::ext_literal]] uint32_t __alignment = alignment);
+
+template<typename T, typename P>
+[[vk::ext_instruction(spv::OpLoad)]]
+enable_if_t<is_spirv_type_v<P>, T> load(P pointer);
+
+template<typename T>
+[[vk::ext_instruction(spv::OpLoad)]]
+T load(pointer_t<spv::StorageClassPhysicalStorageBuffer, T> pointer, [[vk::ext_literal]] uint32_t memoryAccess);
+
+template<typename T>
+[[vk::ext_instruction(spv::OpLoad)]]
+T load(pointer_t<spv::StorageClassPhysicalStorageBuffer, T> pointer, [[vk::ext_literal]] uint32_t memoryAccess, [[vk::ext_literal]] uint32_t memoryAccessParam);
+
+template<typename T, uint32_t alignment>
+[[vk::ext_instruction(spv::OpLoad)]]
+T load(pointer_t<spv::StorageClassPhysicalStorageBuffer, T> pointer, [[vk::ext_literal]] uint32_t __aligned = /*Aligned*/0x00000002, [[vk::ext_literal]] uint32_t __alignment = alignment);
+
+template<typename T>
+[[vk::ext_instruction(spv::OpLoad)]]
+T load(pointer_t<spv::StorageClassPhysicalStorageBuffer, T> pointer);
```

Comment:
This is a BDA load, and as such:
1. it will match the loads above because `pointer_t` is a spirv type, and never get called or worse, clash being ambiguous
2. BDA load/store ALWAYS take aligned operands and require alignment to be specified (read the spec)

Reply:
> it will match the loads above because pointer_t is a spirv type, and never get called or worse, clash being ambiguous

hmm do you have any idea beside renaming the BDA overload to something else?

Reply:
enable_if